### ISSUE TITLE
ACL and Dictionary dependency issue

### ISSUE THREAD

**Issue by tibbon** [fastly/terraform-provider-fastly#309]
### Terraform Version
0.13.2, but also impacts prior versions

### Affected Resource(s)
- fastly_service_acl_entries_v1, entries
- fastly_service_acl_entries_v1, items

### Terraform Configuration Files
```hcl
resource "fastly_service_v1" "aws-staging" {
  name = "aws-staging"

  acl {
    name = var.auto_ban_acl
  }

  dictionary {
    name = var.api_generated_ban_config
  }
  # Many things removed, but this is the relevant snippet above
}

resource "fastly_service_acl_entries_v1" "entries" {
  service_id = fastly_service_v1.aws-staging.id

  # This has been tried via name, and also by just [0] index.
  acl_id = {for x in fastly_service_v1.aws-staging.acl : x.name => x.acl_id}[var.auto_ban_acl]

  # Allow updates from the UI and API without clobbering them
  lifecycle {
    ignore_changes = [entry,]
  }
}

resource "fastly_service_dictionary_items_v1" "items" {
    service_id = fastly_service_v1.aws-staging.id

    # This has been tried via name, and also by just [0] index.
    dictionary_id = {for x in fastly_service_v1.aws-staging.dictionary : x.name => x.dictionary_id}[var.api_generated_ban_config]

    items = {
        enabled: "true"
    }

    # Allow updates from the UI and API without clobbering them
    lifecycle {
      ignore_changes = [items,]
    }
}
```

### Debug Output
A shorter/redacted debug log is here: https://gist.github.com/tibbon/d48cd1e25613e8a104aa31eff557d838

### Expected Behavior
ACL and Dictionary created without error

### Actual Behavior
It seems there's a dependency issue here. The real problem centers around `fastly_service_v1.aws-staging.acl is empty set of object`

```
Error: Invalid index

  on main.tf line 578, in resource "fastly_service_acl_entries_v1" "entries":
 578:   acl_id = {for x in fastly_service_v1.aws-staging.acl : x.name => x.acl_id}[var.auto_ban_acl]
    |----------------
    | fastly_service_v1.aws-staging.acl is empty set of object
    | var.auto_ban_acl is "api_generated_banned_ips"

The given key does not identify an element in this collection value.


Error: Invalid index

  on main.tf line 588, in resource "fastly_service_dictionary_items_v1" "items":
 588:     dictionary_id = {for x in fastly_service_v1.aws-staging.dictionary : x.name => x.dictionary_id}[var.api_generated_ban_config]
    |----------------
    | fastly_service_v1.aws-staging.dictionary is empty set of object
    | var.api_generated_ban_config is "api_generated_ban_config"

The given key does not identify an element in this collection value.
```

### Summary

It seems that if I try to create the `resource "fastly_service_dictionary_items_v1" "items"` or `entries` at the same time as adding it in the `resource "fastly_service_v1" "aws-staging"` then it doesn't yet have the item, so it can't reference it. I've tried to use the new `depends_on` in tf13 without any luck.

I'm following pretty close to the docs below. I can do this as two separate plan/apply steps, but that's no good in a long-term manner.

### References
Following these references:
- https://registry.terraform.io/providers/fastly/fastly/latest/docs/resources/service_acl_entries_v1
- https://registry.terraform.io/providers/fastly/fastly/latest/docs/resources/service_dictionary_items_v1
-

**Comment by smaeda-ks** [comment]
Hi @tibbon,

As we discussed in a separate channel, this is not a bug but a limitation.
You need to create acl/dictionary container first then your `fastly_service_acl_entries_v1`/`fastly_service_dictionary_items_v1 ` resource should work as intended. If the acl/dictionary you specified in your `fastly_service_v1` resource doesn’t exist in your service yet, and you try to plan/apply that creates acl/dictionary container and its associated items at the same time, that error happens.

You can manually create containers via UI/API, or consider using `-target` option to run `fastly_service_v1` resource first then re-apply.

**Comment by tibbon** [comment]
Thanks for some workarounds! The problem I see what that two-phase approach is that it doesn't work well for a fully automated environment that applies all TF changes and can spin up new services/environments. Ideally, I like to think of my Terraform as something that I could rip everything down, run `terraform apply` and have it all back there again.

I got around it by wrapping the `entry_id` assignment in a `try()`, but that felt a bit hacky. I wonder if there's some way in the provider to ensure that it first creates the acl/dictionary, and then applies the updates to the `fastly_service_acl_entries_v1` afterward as a dependency? I tried using `depends_on` to explicitly state this, but it didn't work.

**Comment by phamann** [comment]
Hi all (@tibbon). A very delayed update on issue - which I personally apologise for, as our initial response was incorrect. This issue - and other similar reports - prompted us to do some further analysis into the underlying problem and work out why it is happening. Fortunately, there is a confirmed workaround :tada:

**TL;DR**
We have determined that this is a bug with the way Terraform evaluates _Computed_ attributes inside `Sets`, we have since reported the problem [upstream](https://github.com/hashicorp/terraform/issues/26419) to HashiCorp and have found a workaround which resolves the problem using the new  `for_each` syntax, lastly we have updated our documentation for these resources with the workaround. 

For a detailed explanation see the documentation PR #324 and associated Terraform [issue](https://github.com/hashicorp/terraform/issues/26419). However, please let us know in this thread if the suggested solution resolves your problem or you find any other related issues 😄  

For those interested, here is a more detailed explanation taken from #324:

---

In order to use _versionless_ resources (i.e. ACL Entries, Dictionary Items, Dynamic Snippets Content), it's required to link the ID generated from the Fastly Service (`fastly_service_v1`) to the respective field in the _versionless_ resource (i.e `fastly_service_acl_entries_v1`, `fastly_service_dictionary_items_v1`, `fastly_service_dynamic_snippet_content_v1`).

The original documentation instructed to the use of _for expressions_ to create this link. For example:

``` 
resource "fastly_service_v1" "myservice" {
  ...
  dictionary {
    name = "dict1"
  }

}
resource "fastly_service_dictionary_items_v1" "items" {
  service_id = fastly_service_v1.myservice.id
  dictionary_id = {for d in fastly_service_v1.myservice.dictionary : d.name => d.dictionary_id}["dict1"]
  ...
}
```

This seems to work for many scenarios, however, not for all. 
It fails mainly when the resource `fastly_service_v1` already exists and we add the `acl` block at the same time, we add `fastly_service_acl_entries_v1` resources (The same occur for the other _versionless_ resources).

The generated error is: `Error: Invalid index`

This seems to be due to the way Terraform evaluates *Computed* attributes inside `Sets`. When the resource doesn't exist, Terraform postpones the evaluation of the expression as the values are [not yet known](https://www.terraform.io/docs/configuration/expressions.html%23values-not-yet-known).
However, when the resource exists and the  Set attribute is being modified, Terraform tries to evaluate the expression before the all values are known causing the error. This is likely to be an issue in Terraform.

When replacing the _for expression_ with a `for_each` attribute, Terraform is able to resolve the values correctly. That is, Terraform detects that a value the computed id is still not available and postpones the expression evaluation. 

Interestingly enough, if using the _for expression_ with an `if` and `lookup`, the issue seems to also disappear. For example

```
lookup​({​for​ ​d​ in ​fastly_service_v1​.myservice.dictionary : ​d​.​name​ ​=>​ ​d​.dictionary_id ​if​ ​d​.​name == "dict1"​}, "dict1", ​""​)
```
This approach however is not as clean as the `for_each` one.

For all those reasons, the documentation was updated to recommend the `for_each` approach.

A similar issue also happens to `List` types and current affects the WAF resource. (This issue was previously raised against Terraform - see https://github.com/hashicorp/terraform/issues/23488)

**Comment by phamann** [comment]
@tibbon I'm now closing this given the answer above. Please feel free to open if you're still having issues.

**Comment by tibbon** [comment]
For future readers, I addressed this by creating a `fastly-dictionary` module in my code that I can reuse. The heart of the module code is:

```ruby
resource "fastly_service_dictionary_items_v1" "items" {
  for_each = var.dictionaries

  service_id    = var.service_id
  dictionary_id = { for x in var.service_dictionaries : x.name => x.dictionary_id }[each.key]

  items = each.value

  # Allow updates from the UI and API without clobbering them
  lifecycle {
    ignore_changes = [items, ]
  }

  depends_on = [var.dependencies]
}
```

This is then called via:

```ruby
module "dictionaries" {
  source = "../fastly-dictionary"
  dictionaries = {
    (var.dictionary_1) = {
      flag1 : "true"
    }
    (var.dictionary_2) = {
      flag2 : "false"
      flag3_ttl : "300" # seconds
    }
  }
  service_id                  = fastly_service_v1.configuration.id
  service_dictionaries = fastly_service_v1.configuration.dictionary
  dependencies           = [fastly_service_v1.configuration]
}
```

With this I can easily create more dictionaries, give them default values, not clobber changes made in the UI, etc.
